packages/@lwc/engine-dom/src/custom-elements/create-custom-element-vanilla.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isUndefined } from '@lwc/shared';
+import type { LifecycleCallback } from '@lwc/engine-core';
+
+const cachedConstructors = new Map<string, CustomElementConstructor>();
+const elementsUpgradedOutsideLWC = new WeakSet<HTMLElement>();
+let elementBeingUpgradedByLWC = false;
+
+// Creates a constructor that is intended to be used as a vanilla custom element, except that the upgradeCallback is
+// passed in to the constructor so LWC can reuse the same custom element constructor for multiple components.
+// Another benefit is that only LWC can create components that actually do anything – if you do
+// `customElements.define('x-foo')`, then you don't have access to the upgradeCallback, so it's a dummy custom element.
+// This class should be created once per tag name.
+const createUpgradableConstructor = (
+    connectedCallback: LifecycleCallback,
+    disconnectedCallback: LifecycleCallback
+) => {
+    // TODO [#2972]: this class should expose observedAttributes as necessary
+    return class UpgradableConstructor extends HTMLElement {
+        constructor(upgradeCallback: LifecycleCallback) {
+            super();
+            // If the element is not created using lwc.createElement(), e.g. `document.createElement('x-foo')`,
+            // then elementBeingUpgraded will be false
+            if (elementBeingUpgradedByLWC) {
+                upgradeCallback(this);
+            } else {
+                // keep track of elements that were not created by lwc.createElement,
+                // so we can ignore their lifecycle hooks
+                elementsUpgradedOutsideLWC.add(this);
+
+                // TODO [#2970]: LWC elements cannot be upgraded via new Ctor()
+                // Do we want to support this? Throw an error? Currently for backwards compat it's a no-op.
```

Comment:
After discussion with @ravijayaramappa, I decided to make this a no-op rather than an Error. There is a test that bifurcates (pivot vs non-pivot) to test both behaviors (pivots throw an Error).

In #2970 we have some ongoing discussion on how to handle `new Ctor()` and whether users should be able to instantiate LWC components that way.

I considered adding a `console.error` here, but there are lots of places in the Karma tests where I'd have to add `expect().toLogErrorDev` so I punted on that for now. We can always add a `console.error` later.